### PR TITLE
Generate completion code

### PR DESCRIPTION
**Pull request by nlewo**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -794,10 +794,18 @@ func main() {
 
 #### Enabling
 
-Source the `autocomplete/bash_autocomplete` file in your `.bashrc` file while
-setting the `PROG` variable to the name of your program:
+You can generate bash or zsh completion code by using the flag `--init-completion bash` or `--init-completion bash`.
```
Comment (meatballhat) [inline README.md]: This should mention `--init-completion zsh`, right? :smile_cat:

File: flag.go [hunk 2]
```diff
@@ -21,12 +21,18 @@ var (
...
+var GenerateCompletionFlag = &BoolFlag{
+	Name:   "generate-completion",
 	Hidden: true,
 }
 
+// InitCompletionFlag generates completion code
+var InitCompletionFlag = &StringFlag{
+	Name:  "init-completion",
+	Usage: "generate completion code. Value must be 'bash' or 'zsh'",
+}
```
Comment (meatballhat) [inline flag.go]: This makes me think that it'd be nice to have completion code that's extensible from outside this package.  No need to try to do that now, though...

**Comment by meatballhat** [comment]
@nlewo I think this is on the right track :+1:.

- Do you know why the generated flag code is being deleted?
- Are you able to take a stab at migration code so that tests for that pass?

File: flag.go (hunk 2 above)
Comment (nlewo) [inline flag.go]: We can consider the current completion code as default completion code. Another flag could then be added later to provide custom completion code.

File: README.md (hunk 1 above)
Comment (nlewo) [inline README.md]: Yes, fixed.

**Comment by nlewo** [comment]
Concerning your first point, it was due to a rebase... Fixed.

**Comment by nlewo** [comment]
@meatballhat ok. Thanks

**Comment by meatballhat** [comment]
@nlewo I'm happy to help with the migration stuff, fwiw.  Just lemme know :+1: :smile_cat:

**Comment by nlewo** [comment]
@meatballhat I'm not able to run migration test suite. I get the following error:
```
./runtests migrations
runtests: curl -sSL -o README.md https://raw.githubusercontent.com/urfave/cli/v1/README.md
runtests: gfmrun extract -o .
WARN[0000] no sources given 
```

**Comment by meatballhat** [comment]
@nlewo the gfmrun executable was susceptible to a recent context lookup bug in v2, so you will need to pull latest in order to get it working again.

**Comment by nlewo** [comment]
@meatballhat I provided migration functions. Ok, they are really fragile... but tests are ok and it works on my app :/

File: cli-v1-to-v2 [hunk 3]
```diff
@@ -298,6 +298,30 @@ def _app_init(source):
...
+@_migrator
+def _bash_completion_flag(source):
+    return re.sub(
+        'cli\\.BashCompletionFlag', 'cli.GenerateCompletionFlag', source, flags=re.UNICODE
+    )
+
+@_migrator
+def _args(source):
+    return re.sub(
+        '\\.Args():', '.Args.Slice()', source, flags=re.UNICODE
```
Comment (meatballhat) [inline cli-v1-to-v2]: Shouldn't this be '`\\.Args\(\)` (escaped and without the colon)?

**Comment by meatballhat** [comment]
@nlewo sorry for the slow feedback.  Just the one in-line question/comment.

File: cli-v1-to-v2 (hunk 3 above)
Comment (nlewo) [inline cli-v1-to-v2]: Sorry. Fixed.

**Comment by nlewo** [comment]
@meatballhat I removed the migration function for the Args pattern since it is not related to the completion code. I think I unintentionally added it in a commit since it is needed to test the migration on my app.

**Comment by nlewo** [comment]
@meatballhat Could you have a look please?

**Comment by meatballhat** [comment]
@nlewo LGTM :shipit: 

/cc @jszwedko feedback?

**Comment by jszwedko** [comment]
This is pretty nifty, I like this :+1: :+1:

**Comment by nlewo** [comment]
@jszwedko Ok. Cool. 
@meatballhat could you merge it? ;)

**Comment by nlewo** [comment]
@jszwedko @meatballhat any news on this PR?

**Comment by jszwedko** [comment]
Apologies for the delay @nlewo -- checked with @meatballhat out-of-band for confirmation -- merging now. Thank you so much for putting this together and cooperating through the changes!

**Comment by nlewo** [comment]
@jszwedko I have more pull requests concerning the completion, but I'm quite afraid by pushing them regarding the delay:/

**Comment by jszwedko** [comment]
@nlewo I can understand that and I do apologize for the delay in the merge; this month has also been a busy one for me. I have some more free time now and can promise to be more responsive if you have more changes you'd like to push.

**Comment by imle** [comment]
This seems to have disappeared from the codebase, is there any reason it was removed?

**Comment by lynncyrin** [comment]
@imle Can you please provide a citation for:

> This seems to have disappeared from the codebase

Because I doubt anyone is going to know offhand what you're referring to.

**Pull request merged**